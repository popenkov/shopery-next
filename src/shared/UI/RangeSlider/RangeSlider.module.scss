@import '@/shared/styles/breakpoints';
@import '@/shared/styles/mixins/text-styles.scss';
@import '@/shared/styles/mixins/box.scss';

.range {
  padding: 16px 0 8px;
}

.text {
  @include body-s;

  color: var(--gray-700);
}

.textValue {
  color: var(--gray-900);
  font-weight: 500;
}

.slider {
  height: 6px;
  background-color: var(--gray-100);
  border-radius: 15px;
}

.slider + .text {
  display: block;
  margin-top: 22px;
}

.rangeSlider {
  :global(.rc-slider-rail) {
    height: 6px;
    background-color: var(--gray-100);
    border-radius: 15px;
  }

  :global(.rc-slider-track) {
    height: 6px;
    background-color: var(--primary);
    border-radius: 15px;
  }

  :global(.rc-slider-handle) {
    @include box(18px);

    background-color: var(--back-primary);
    border: 2px solid var(--primary);
    border-radius: 15px;

    @media (hover: hover) {
      &:hover {
        background-color: var(--primary);
      }
    }
  }
}
