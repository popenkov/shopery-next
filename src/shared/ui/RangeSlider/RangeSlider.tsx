'use client';

import {
    DetailedHTMLProps,
    FC,
    HTMLAttributes,
    memo,
    useEffect,
    useState,
} from 'react';

import Slider from 'rc-slider';
import 'rc-slider/assets/index.css';

import cls from './RangeSlider.module.scss';

interface RangeValueInterface {
    min: number;
    max: number;
}

interface Props
    extends Omit<
            DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>,
            'onChange'
        >,
        RangeValueInterface {
    range: number[];
    onChange: (value: RangeValueInterface) => void;
}

export const RangeSlider: FC<Props> = memo((props) => {
    const { min, max, range, onChange } = props;
    const [currentRange, setCurrentRange] = useState<number[]>(range);

    const handleSliderUpdate = (value: number | number[]) => {
        if (Array.isArray(value)) {
            const [min, max] = value;
            setCurrentRange(value);
            onChange({ min, max });
        }
    };
    return (
        <div className={cls.range}>
            <Slider
                range
                min={min}
                max={max}
                defaultValue={range}
                value={currentRange}
                onChange={handleSliderUpdate}
                className={cls.rangeSlider}
            />
            <span className={cls.text}>
                Price:{' '}
                <span className={cls.textValue}>
                    {currentRange[0]} - {currentRange[1]}
                </span>
            </span>
        </div>
    );
});

RangeSlider.displayName = 'RangeSlider';
