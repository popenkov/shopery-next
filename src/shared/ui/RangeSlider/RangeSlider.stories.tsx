import type { Meta, StoryObj } from '@storybook/react';

import { RangeSlider } from './RangeSlider';

const meta = {
  title: 'shared/RangeSlider',
  component: RangeSlider,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
} satisfies Meta<typeof RangeSlider>;

export default meta;
type TStory = StoryObj<typeof meta>;

export const RangeSliderComponent: TStory = {
  args: {
    min: 0,
    max: 550,
    range: [0, 550],
    onChange: () => {
      console.log('change');
    },
  },
};
