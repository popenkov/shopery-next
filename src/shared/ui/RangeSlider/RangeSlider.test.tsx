import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import { RangeSlider } from './RangeSlider';

describe('RangeSlider component', () => {
  it('renders correctly', () => {
    const { container } = render(
      <RangeSlider min={0} max={100} range={[20, 80]} onChange={() => {}} />,
    );
    expect(container).toMatchSnapshot();
  });

  // todo разобраться как изменить значение

  // it('calls onChange callback when slider is updated', () => {
  //   const onChange = jest.fn();
  //   const { getByRole, container } = render(
  //     <RangeSlider min={0} max={100} range={[20, 80]} onChange={onChange} />,
  //   );
  //   const slider = container.querySelector('.rangeSlider');
  //   fireEvent.change(container, { target: { value: [30, 90] } });
  //   expect(onChange).toHaveBeenCalledTimes(1);
  //   expect(onChange).toHaveBeenCalledWith({ min: 30, max: 90 });
  // });

  // it('updates currentRange state when slider is updated', () => {
  //   const { getByRole, getByText } = render(
  //     <RangeSlider min={0} max={100} range={[20, 80]} onChange={() => {}} />,
  //   );
  //   const slider = getByRole('slider');
  //   fireEvent.change(slider, { target: { value: [30, 90] } });
  //   const textValue = getByText('30 - 90');
  //   expect(textValue).toBeInTheDocument();
  // });

  it('renders correct text value when slider is updated', () => {
    const { getByText } = render(
      <RangeSlider min={0} max={100} range={[20, 80]} onChange={() => {}} />,
    );
    const textValue = getByText('20 - 80');
    expect(textValue).toBeInTheDocument();
  });

  it('has correct class names', () => {
    const { container } = render(
      <RangeSlider min={0} max={100} range={[20, 80]} onChange={() => {}} />,
    );
    expect(container.querySelector('.range')).toBeInTheDocument();
    expect(container.querySelector('.rangeSlider')).toBeInTheDocument();
    expect(container.querySelector('.text')).toBeInTheDocument();
    expect(container.querySelector('.textValue')).toBeInTheDocument();
  });
});
