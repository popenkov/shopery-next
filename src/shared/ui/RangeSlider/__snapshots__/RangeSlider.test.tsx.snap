// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeSlider component renders correctly 1`] = `
<div>
  <div
    class="range"
  >
    <div
      class="rc-slider rangeSlider rc-slider-horizontal"
    >
      <div
        class="rc-slider-rail"
      />
      <div
        class="rc-slider-track rc-slider-track-1"
        style="left: 20%; width: 60%;"
      />
      <div
        class="rc-slider-step"
      />
      <div
        aria-disabled="false"
        aria-orientation="horizontal"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="20"
        class="rc-slider-handle rc-slider-handle-1"
        role="slider"
        style="left: 20%; transform: translateX(-50%);"
        tabindex="0"
      />
      <div
        aria-disabled="false"
        aria-orientation="horizontal"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="80"
        class="rc-slider-handle rc-slider-handle-2"
        role="slider"
        style="left: 80%; transform: translateX(-50%);"
        tabindex="0"
      />
    </div>
    <span
      class="text"
    >
      Price:
       
      <span
        class="textValue"
      >
        20
         - 
        80
      </span>
    </span>
  </div>
</div>
`;
